File: LinkedList/python/problem_061_rotate_list.py
"""
Problem 61: Rotate List
Difficulty: Medium

Problem Statement:
Given the head of a linked list, rotate the list to the right by k places.

Example:
Input: head = [1,2,3,4,5], k = 2
Output: [4,5,1,2,3]

Approach:
1. Find length of list
2. Connect tail to head (make it circular)
3. Find new tail position: (length - k % length - 1)
4. Break the circle at new tail

Time Complexity: O(n)
Space Complexity: O(1)
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head or not head.next or k == 0:
            return head
        
        # Step 1: Find length and get to tail
        tail = head
        length = 1
        while tail.next:
            tail = tail.next
            length += 1
        
        # Step 2: Make it circular
        tail.next = head
        
        # Step 3: Find new tail position
        # After rotation, new tail is at position (length - k % length - 1)
        k = k % length  # Handle k > length
        steps_to_new_tail = length - k - 1
        
        new_tail = head
        for _ in range(steps_to_new_tail):
            new_tail = new_tail.next
        
        # Step 4: Break the circle
        new_head = new_tail.next
        new_tail.next = None
        
        return new_head

def create_list(arr):
    if not arr:
        return None
    head = ListNode(arr[0])
    current = head
    for i in range(1, len(arr)):
        current.next = ListNode(arr[i])
        current = current.next
    return head

def print_list(head):
    result = []
    current = head
    while current:
        result.append(str(current.val))
        current = current.next
    print("[" + ",".join(result) + "]")

if __name__ == "__main__":
    solution = Solution()
    
    head = create_list([1, 2, 3, 4, 5])
    print("Original: ", end="")
    print_list(head)
    
    result = solution.rotateRight(head, 2)
    print("After rotating by 2: ", end="")
    print_list(result)  # Output: [4,5,1,2,3]

File: LinkedList/java/Problem061_RotateList.java
/**
 * Problem 61: Rotate List
 * Difficulty: Medium
 * 
 * Problem Statement:
 * Given the head of a linked list, rotate the list to the right by k places.
 * 
 * Example:
 * Input: head = [1,2,3,4,5], k = 2
 * Output: [4,5,1,2,3]
 * 
 * Approach:
 * 1. Find length of list
 * 2. Connect tail to head (make it circular)
 * 3. Find new tail position: (length - k % length - 1)
 * 4. Break the circle at new tail
 * 
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

public class Problem061_RotateList {
    
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null || k == 0) {
            return head;
        }
        
        // Step 1: Find length and get to tail
        ListNode tail = head;
        int length = 1;
        while (tail.next != null) {
            tail = tail.next;
            length++;
        }
        
        // Step 2: Make it circular
        tail.next = head;
        
        // Step 3: Find new tail position
        // After rotation, new tail is at position (length - k % length - 1)
        k = k % length;  // Handle k > length
        int stepsToNewTail = length - k - 1;
        
        ListNode newTail = head;
        for (int i = 0; i < stepsToNewTail; i++) {
            newTail = newTail.next;
        }
        
        // Step 4: Break the circle
        ListNode newHead = newTail.next;
        newTail.next = null;
        
        return newHead;
    }
    
    public static ListNode createList(int[] arr) {
        if (arr.length == 0) return null;
        ListNode head = new ListNode(arr[0]);
        ListNode current = head;
        for (int i = 1; i < arr.length; i++) {
            current.next = new ListNode(arr[i]);
            current = current.next;
        }
        return head;
    }
    
    public static void printList(ListNode head) {
        ListNode current = head;
        System.out.print("[");
        while (current != null) {
            System.out.print(current.val);
            if (current.next != null) System.out.print(",");
            current = current.next;
        }
        System.out.println("]");
    }
    
    public static void main(String[] args) {
        Problem061_RotateList solution = new Problem061_RotateList();
        
        ListNode head = createList(new int[]{1, 2, 3, 4, 5});
        System.out.print("Original: ");
        printList(head);
        
        ListNode result = solution.rotateRight(head, 2);
        System.out.print("After rotating by 2: ");
        printList(result);  // Output: [4,5,1,2,3]
    }
}
